x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-BlockBuilder.java.st

```diff
@@ -281,26 +281,17 @@ $if(BytesRef)$
         values = null;
 $else$
         if (hasNonNullValue && positionCount == 1 && valueCount == 1) {
-            block = new Constant$Type$Vector(values[0], 1, blockFactory).asBlock();
+            block = blockFactory.newConstant$Type$BlockWith(values[0], 1, estimatedBytes);
         } else {
             if (values.length - valueCount > 1024 || valueCount < (values.length / 2)) {
                 values = Arrays.copyOf(values, valueCount);
             }
             if (isDense() && singleValued()) {
-                block = new $Type$ArrayVector(values, positionCount, blockFactory).asBlock();
+                block = blockFactory.new$Type$ArrayVector(values, positionCount, estimatedBytes).asBlock();
             } else {
-                block = new $Type$ArrayBlock(values, positionCount, firstValueIndexes, nullsMask, mvOrdering, blockFactory);
+                block = blockFactory.new$Type$ArrayBlock(values, positionCount, firstValueIndexes, nullsMask, mvOrdering, estimatedBytes);
             }
         }
-        /*
-         * Update the breaker with the actual bytes used.
-         * We pass false below even though we've used the bytes. That's weird,
-         * but if we break here we will throw away the used memory, letting
-         * it be deallocated. The exception will bubble up and the builder will
-         * still technically be open, meaning the calling code should close it
-         * which will return all used memory to the breaker.
-         */
-        blockFactory.adjustBreaker(block.ramBytesUsed() - estimatedBytes, false);
```

Comment:
Shouldn't we also remove the memory accounting from the ctor?

Reply:
I don't think so. The general idea here is to seed the initial memory adjustment based on the estimated size, then allow the memory to be adjusted incrementally as the builder has values added. The initial adjustment is done through `AbstractBlockBuilder::adjustBreaker`. Is it that the naked name `adjustBreaker` in the constructor is confusing? Or something else?

Reply:
Uhm... yes, I see, you are also passing it to the factory as `preAdjustedBytes`. It's a bit confusing, but I think it's also fragile, eg. if the breaker fails, the builder remains open and the pre-allocated value is never released. 
Also, the factory has a problem with `adjustBreaker()`, see [this](https://github.com/elastic/elasticsearch/pull/100136), that makes the whole logic even more obscure IMHO

Reply:
No, probably that's correct, but this mixed/shared manipulation of the circuit breaker is definitely confusing